File: 백준/Gold/7576. 토마토/토마토.py
from collections import deque

m, n = map(int, input().split())
lst = [list(map(int, input().split())) for _ in range(n)]
dist = [(1, 0), (-1, 0), (0, 1), (0, -1)]


def bfs(n, m, array):
    q = deque()
    v = [[False] * m for _ in range(n)]
    for i in range(n):
        for j in range(m):
            if lst[i][j] == 1:
                q.append((i, j))
                v[i][j] = True
    while q:
        x, y = q.popleft()
        for dx, dy in dist:
            nx = x + dx
            ny = y + dy
            if 0 <= nx < n and 0 <= ny < m and not v[nx][ny] and array[nx][ny] == 0:
                v[nx][ny] = True
                array[nx][ny] = array[x][y] + 1
                q.append((nx, ny))


bfs(n, m, lst)
max_day = 0
done = True

for i in range(n):
    for j in range(m):
        if lst[i][j] == 0:
            done = False
        max_day = max(max_day, lst[i][j])

print(max_day - 1 if done else -1)

File: 백준/Gold/7576. 토마토/토마토.cc
#include <iostream>
#include <vector>
#include <queue>
#include <algorithm>

using namespace std;

int m, n;
vector<vector<int>> lst;
vector<pair<int, int>> dist = {
	{1, 0}, {-1, 0}, {0, 1}, {0, -1}
};

void bfs() {
	queue<pair<int, int>> q;
	vector<vector<bool>> v(n, vector<bool>(m, false));

	for (int i = 0; i < n; i++) {
		for (int j = 0; j < m; j++) {
			if (lst[i][j] == 1) {
				q.push({ i, j });
				v[i][j] = true;
			}
		}
	}
	while (!q.empty()) {
		pair<int, int> pos = q.front();
		int x = pos.first;
		int y = pos.second;
		q.pop();

		for (const pair<int, int> dir : dist) {
			int dx = dir.first;
			int dy = dir.second;
			int nx = x + dx;
			int ny = y + dy;
			if (0 <= nx && nx < n && 0 <= ny && ny < m && !v[nx][ny] && lst[nx][ny] == 0) {
				v[nx][ny] = true;
				lst[nx][ny] = lst[x][y] + 1;
				q.push({ nx, ny });
			}
		}
	}
}

int main() {
	cin >> m >> n;
	lst.resize(n, vector<int>(m));
	
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < m; j++) {
			cin >> lst[i][j];
		}
	}

	bfs();
	int max_day = 0;
	bool flag = true;
	for (int i = 0; i < n; i++) {
		for (int j = 0; j < m; j++) {
			if (lst[i][j] == 0) flag = false;
			max_day = max(max_day, lst[i][j]);
		}
	}

	if (flag) cout << max_day - 1;
	else cout << -1;
}